src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexQueryPlan.java

```diff
@@ -0,0 +1,135 @@
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableSet;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.ReadCommand;
+import org.apache.cassandra.db.filter.RowFilter;
+import org.apache.cassandra.index.Index;
+import org.apache.cassandra.index.sai.StorageAttachedIndex;
+import org.apache.cassandra.index.sai.metrics.TableQueryMetrics;
+import org.apache.cassandra.schema.TableMetadata;
+
+public class StorageAttachedIndexQueryPlan implements Index.QueryPlan
+{
+    private final ColumnFamilyStore cfs;
+    private final TableQueryMetrics queryMetrics;
+    private final RowFilter postIndexFilter;
+    private final RowFilter filterOperation;
+    private final Set<Index> indexes;
+
+    private StorageAttachedIndexQueryPlan(ColumnFamilyStore cfs,
+                                          TableQueryMetrics queryMetrics,
+                                          RowFilter postIndexFilter,
+                                          RowFilter filterOperation,
+                                          ImmutableSet<Index> indexes)
+    {
+        this.cfs = cfs;
+        this.queryMetrics = queryMetrics;
+        this.postIndexFilter = postIndexFilter;
+        this.filterOperation = filterOperation;
+        this.indexes = indexes;
+    }
+
+    @Nullable
+    public static StorageAttachedIndexQueryPlan create(ColumnFamilyStore cfs,
+                                                       TableQueryMetrics queryMetrics,
+                                                       Set<StorageAttachedIndex> indexes,
+                                                       RowFilter rowFilter)
+    {
+        ImmutableSet.Builder<Index> selectedIndexesBuilder = ImmutableSet.builder();
+
+        for (RowFilter.Expression expression : rowFilter)
+        {
+            // we ignore user-defined expressions here because we don't have a way to translate their #isSatifiedBy
+            // method, they will be included in the filter returned by QueryPlan#postIndexQueryFilter()
+            if (expression.isUserDefined())
+                continue;
+
+            for (StorageAttachedIndex index : indexes)
+            {
+                if (index.supportsExpression(expression.column(), expression.operator()))
+                {
+                    selectedIndexesBuilder.add(index);
+                }
+            }
+        }
+
+        ImmutableSet<Index> selectedIndexes = selectedIndexesBuilder.build();
+        if (selectedIndexes.isEmpty())
+            return null;
+
+        /*
+         * postIndexFilter comprised by those expressions in the read command row filter that can't be handled by
+         * {@link FilterTree#satisfiedBy(Unfiltered, Row, boolean)}. That includes expressions targeted
+         * at {@link RowFilter.UserExpression}s like those used by RLAC.
```

Comment:
nit: What's RLAC? ;)

Reply:
I have tried to remove comments like this and keep looking for them. I'll get rid of this.

Reply:
I have tried to remove comments like this and keep looking for them. I'll get rid of this.